File: Meli/LeetCode 75/450. Delete Node in a BST.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if not root:
            return None
        
        if key < root.val:
            root.left = self.deleteNode(root.left, key)
        elif key > root.val:
            root.right = self.deleteNode(root.right, key)
        else:
            # Case 1: Leaf node
            if not root.left and not root.right:
                return None
            # Case 2: Node with only one child
            elif not root.left:
                return root.right
            elif not root.right:
                return root.left
            # Case 3: Node with two children
            else:
                # Find the minimum value in right subtree
                min_node = self.findMin(root.right)
                root.val = min_node.val
                root.right = self.deleteNode(root.right, min_node.val)
        
        return root
    
    def findMin(self, node: TreeNode) -> TreeNode:
        current = node
        while current.left:
            current = current.left
        return current

File: Gombang/[LeetCode75] - [Binary_Search_Tree]/450_Delete_Node_in_a_BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    TreeNode* findMaxNode(TreeNode* root)
    {
        if (root->right == nullptr)
            return root;

        return findMaxNode(root->right);
    }

    TreeNode* deleteNode(TreeNode* root, int key) {

        if (root == nullptr)
            return nullptr;

        if (root->val < key)
        {
            root->right = deleteNode(root->right, key);
        }
        else if (root->val > key)
        {
            root->left = deleteNode(root->left, key);
        }
        else
        {
            if (root->left == nullptr)
                return root->right;

            if (root->right == nullptr)
                return root->left;

            // 왼쪽 서브 트리에서 가장 큰 값을 찾는다.
            TreeNode* maxNode = findMaxNode(root->left);
            root->val = maxNode->val;
            root->left = deleteNode(root->left, maxNode->val);
        }

        return root;
    }
};